src/feature/hs/hs_common.c

```diff
@@ -1723,45 +1714,45 @@ hs_get_extend_info_from_lspecs(const smartlist_t *lspecs,
     }
   } SMARTLIST_FOREACH_END(ls);
 
+  fascist_firewall_choose_address_ls(lspecs, FIREWALL_OR_CONNECTION, 0, &ap);
+
   /* Legacy ID is mandatory, and we require IPv4. */
-  if (!have_v4 || !have_legacy_id) {
+  if (!tor_addr_port_is_valid_ap(&ap, 0) || !have_legacy_id) {
     goto done;
   }
 
-  /* We know we have IPv4, because we just checked. */
+  /* We know we have an IP address, because we just checked. */
   if (!direct_conn) {
-    /* All clients can extend to any IPv4 via a 3-hop path. */
+    /* All clients can extend to any IP via a 3-hop path. */
```

Comment:
Relays won't extend via IPv6, we can only connect via IPv6 when we are connecting directly.